go/pkg/pwagent/daemon.go

```diff
@@ -56,28 +64,121 @@ func Daemon(ctx context.Context, cli *client.Client, logger *zap.Logger) error {
+	for {
+		// fetch local info from docker daemon
+		pathwarInfo, err := pwcompose.GetPathwarInfo(ctx, cli)
+		if err != nil {
+			return errcode.ErrComposeGetPathwarInfo.Wrap(err)
+		}
+		fmt.Println("info", godev.PrettyJSON(pathwarInfo))
 
-	// compute the difference
+		agentOpts := AgentOpts{
+			DomainSuffix:      "local",
+			HostIP:            "0.0.0.0",
+			HostPort:          "8000",
+			ModeratorPassword: "",
+			Salt:              "",
+			AllowedUsers:      map[string][]int64{},
+			ForceRecreate:     false,
+			NginxDockerImage:  "docker.io/library/nginx:stable-alpine",
+		}
+		// compute the difference
+		for _, apiInstance := range apiInstances.GetInstances() {
+			found := false
+			needRedump := false
+			for _, flavor := range pathwarInfo.RunningFlavors {
+				apiInstanceFlavor := apiInstance.GetFlavor()
+				if apiInstanceFlavor != nil {
+					apiInstanceFlavorChallenge := apiInstanceFlavor.GetChallenge()
+					if apiInstanceFlavorChallenge != nil {
+						if flavor.Name == apiInstanceFlavorChallenge.GetName() && flavor.Version == apiInstanceFlavor.GetVersion() {
+							found = true
+							if apiInstance.GetStatus() == pwdb.ChallengeInstance_NeedRedump {
+								needRedump = true
+							}
+						}
+					}
+					seasonChallenges := apiInstanceFlavor.GetSeasonChallenges()
+					if seasonChallenges != nil {
+						for _, seasonChallenge := range seasonChallenges {
+							subscriptions := seasonChallenge.GetSubscriptions()
+							if subscriptions != nil {
+								for _, subscription := range subscriptions {
+									if subscription.GetStatus() == pwdb.ChallengeSubscription_Active {
+										team := subscription.GetTeam()
+										if team != nil {
+											members := team.GetMembers()
+											if members != nil {
+												for _, member := range members {
+													for _, instance := range flavor.Instances {
+														for _, port := range instance.Ports {
+															if port.PublicPort != 0 {
+																// configure nginx
+																// - compute the list of members in Active teams
+																// - generate a hash based on their userIDs
+																instanceName := instance.Names[0][1:]
+																_, entryFound := agentOpts.AllowedUsers[instanceName]
+																if !entryFound {
+																	agentOpts.AllowedUsers[instanceName] = []int64{member.GetID()}
+																} else {
+																	allowedUsersSlice := agentOpts.AllowedUsers[instanceName]
+																	allowedUsersSlice = append(allowedUsersSlice, member.GetID())
+																	agentOpts.AllowedUsers[instanceName] = allowedUsersSlice
+																}
+															}
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
 
-	// start missing instances
+			// start missing instances
+			// redump instances with state=NeedRedump
+			if !found || needRedump {
```

Comment:
To improve code readability (and probably to allow some optimizations)

1. One that constructs a slice of found/needRedump IDs
2. One that iterates over this slice to apply some starting/restarting actions